gap/PolygonalComplexes/graphs.gi

```diff
@@ -1115,4 +1115,337 @@ InstallMethod(AllSimplicialSurfacesOfDigraph,"for a digraph and a Boolean",
+        __DigraphAddEdgeNC(D, numVertices + 1, numVertices + 2);
+        __DigraphAddEdgeNC(D, numVertices + 2, numVertices + 1);
+
+        return D;
+    end
+);
+
+InstallMethod( ButterflyInsertion, "for a digraph and two lists", [IsDigraph, IsList, IsList],
+    function (D, edgeA, edgeB)
+        local vertices, numVertices, vertex, edge, inDegrees;
+
+        if IsEmptyDigraph(D) then
+            return ErrorNoReturn("digraph D must not be empty");
+        fi;
+
+        vertices := Concatenation(edgeA, edgeB);
+        numVertices := Maximum(DigraphVertices(D));
+
+        # Sanity checks
+        if Length(vertices) < 3 then
+            return ErrorNoReturn("edgeA and edgeB both need to contain exactly two vertices");
+        fi;
+        #
+        #
+        if Length(edgeA) <> 2 then
+                return ErrorNoReturn("edgeA has to have a length of exactly 2");
+        fi;
+        if Length(edgeB) <> 2 then
+            return ErrorNoReturn("edgeB has to have a length of exactly 2");
+        fi;
+        #
+        if Maximum(edgeA) > numVertices then
+            return ErrorNoReturn("edgeA contains a vertex index that is not included in the provided graph");
+        fi;
+        if Maximum(edgeB) > numVertices then
+            return ErrorNoReturn("edgeB contains a vertex index that is not included in the provided graph");
+        fi;
+
+        # inDegrees := InDegrees(D);
+        # if Length(inDegrees) <> 1 or inDegrees[1] <> 3 then
+        #     return ErrorNoReturn("Every vertex of D must have exactly three neighbours");
+        # fi;
+
+        D := DigraphSymmetricClosure(D);
+
+        return ButterflyInsertionNC(D, edgeA, edgeB);
+    end
+);
+
+InstallMethod( ButterflyDeletionNC, "for a digraph and a list", [IsDigraph, IsList],
+    function (D, intersectingEdge)
+        local dMutable, isMutable, numVertices;
+
+        if IsMutableDigraph(D) then
+            isMutable := true;
+            dMutable := D;
+        else
+            isMutable := false;
+            dMutable := DigraphMutableCopy(D);
+        fi;
+
+        dMutable := __ButterflyDeletionDirectNC(dMutable, intersectingEdge);
+
+        if isMutable then
+            return dMutable;
+        fi;
+        return DigraphImmutableCopy(dMutable);
+    end
+);
+
+InstallMethod( __ButterflyDeletionDirectNC, "for a mutable digraph and a list", [IsMutableDigraph, IsList],
+    function (D, intersectingEdge)
+        local neighbours, neighbour, newEdge;
+
+        # Add new edges
+        #
+        # left side of intersecting edge
+        newEdge := [];
+        neighbours := OutNeighboursOfVertex(D, intersectingEdge[1]);
+        for neighbour in neighbours do
+            if neighbour <> intersectingEdge[2] then
+                Add(newEdge, neighbour);
+            fi;
+        od;
+        if Length(newEdge) <> 2 then
+            return ErrorNoReturn("intersectingEdge[1] does not have three outwards neighbours as expected");
```

Comment:
This is a no check method so you don't need to check this. You must check in the normal version that the vertices of the edge which you want to delete don't have a common neighbour.